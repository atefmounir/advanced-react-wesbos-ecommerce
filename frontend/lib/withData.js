import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';

import { endpoint,prodEndpoint } from '../config';
import {LOCAL_STATE_QUERY} from "../components/Cart";

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState:{                                                              //setting local data
      resolvers:{                                                              //-resolvers
        Mutation:{                                                             //-as Mutations in backend
          toggleCart(_,variables,{cache}){                                     //-toggleCart Mutation function in Cart.js
            const {cartOpen}=cache.readQuery({                                 //-read the value of state from the cache. it is returned from the Query "data.cartOpen"
              query:LOCAL_STATE_QUERY                                          //-define the Query operation responsible for reading the state
            })

            const setCartOpen={                                                //-define the operation will be done on the state. in our case, we need to toggle the state
              data:{cartOpen:!cartOpen}                                        //-store the changed state on data object
            }
            cache.writeData(setCartOpen)                                       //-use writeData to modify the cache with the new state value

            return setCartOpen                                                 //-return the new state from the mutation function
          }
        }
      },
      defaults:{                                                               //-default states. considered as initial states in redux reducers and hooks
        cartOpen:false,                                                        //--state will be stored in the cache.
      }
    }
  });
}

export default withApollo(createClient);


/*
  Notes
  This configuration used to let React Next.JS can deal with Apollo for Query, Mutation operations and managing local state
  withData used with Apollo Provider in _app.js to create Apollo Client

*/
