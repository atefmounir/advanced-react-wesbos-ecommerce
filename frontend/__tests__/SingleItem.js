import React from 'react'
import {mount} from 'enzyme'                              //to allow component deep rendering
import toJSON from "enzyme-to-json";                      //to convert react component to html
import wait from 'waait'                                  //returns a promise that resolves after a pre-defined time
import {MockedProvider} from 'react-apollo/test-utils'    //for mocking Apollo-client while testing in operation

import SingleItem,{SINGLE_ITEM_QUERY} from "../components/SingleItem";
import {fakeItem} from "../lib/testUtils";


describe("<SingleItem/>",()=>{
  it("renders with proper data", async ()=>{
    const mocks=[                                                             //define the mocks array
      {
        request:{query:SINGLE_ITEM_QUERY, variables:{id:'abc123'}},           //-mocking the query request: give query name and variables
        result:{
          data:{
            item:fakeItem(),
          }
        }
      }
    ]

    const wrapper=mount(                                                      //mocking the Apollo-Provider by wrapping the component with MockedProvider
      <MockedProvider mocks={mocks}>
        <SingleItem id='abc123'/>
      </MockedProvider>
    )

    expect(wrapper.text()).toContain('Loading...')
    await wait()
    wrapper.update()
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot()            //get a snapshot from the wrapper of only h2 tag element
  })

  it('Errors with a not defined item', async () => {
    const mocks=[{
      request:{query:SINGLE_ITEM_QUERY, variables:{id:'abc123'}},             //-mocking the query request: give query name and variables
      result:{
        errors:[{message:'Item not found'}],
      }
    }]

    const wrapper=mount(                                                      //mocking the Apollo-Provider by wrapping the component with MockedProvider
      <MockedProvider mocks={mocks}>
        <SingleItem id='abc123'/>
      </MockedProvider>
    )

    await wait()
    wrapper.update()
    const item=wrapper.find('[data-test="graphql-error"]')                    //check wrapper.debug()
    console.log(item.debug())
    expect(item.text()).toContain('Shoot!Item not found')
    expect(toJSON(item)).toMatchSnapshot()
  })
})

/*
  Notes
  The SingleItem is wrapped via Query component and to get rendered properly, it has to as its parent which is Apollo-Provider
  Since Apollo-Provider is taking off during the testing, we have to mock it with MockedProvider.
  We have to define the mocks array to simulate the query request and results
*/

