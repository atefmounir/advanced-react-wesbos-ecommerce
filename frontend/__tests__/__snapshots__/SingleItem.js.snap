// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem/> Errors with a not defined item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Item not found
</p>
`;

exports[`<SingleItem/> renders with proper data 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;
